[MBQL lib] Cache round trips between legacy and lib in both directions

bshepherdson opened the pull request:
### Description


Previously, conversions from legacy to MLv2 were cached based on the identity of the incoming query.

This PR adds new caching for round trips in both directions:
1. When converting to MLv2, the legacy form is saved in the Clojure metadata. When converting back that value is used
   unless the query has been edited.
2. When converting to legacy, the MLv2 form is saved in a non-enumerable JS property. When converting back to MLv2,
   that value is used unless the `metadata-provider` has changed.

See the comments in `metabase.lib.js/query`  and `legacy-query` for more details.

**NOTE** All of this refers only to the CLJS side.

### How to verify

This should be invisible, except for the speedup. Put a breakpoint in the inner conversion functions like
`legacy-query*` to check if they're called on a round trip.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

bshepherdson committed:
[MBQL lib] Cache round trips between legacy and lib in both directions

See the comments in `metabase.lib.js` about the various approaches to
caching.

bshepherdson commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/45865?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#45865** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45865?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#45809** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45809?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bshepherdson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

ranquild reviewed (approved):
Awesome!

snoe reviewed (approved):
LGTM - Slightly worried that making it easier to use legacy queries removes the impetus to move forward with pmbql in the app.

lbrdnk commented on test/metabase/lib/js_test.cljs:
```diff
@@ -680,3 +680,119 @@
                         :value   nil}]
       (is (js= expected (lib.js/display-info->js input))))))
+
+(deftest ^:parallel fresh-legacy-query-caching-test
```
This test (and others) is marked parallel while using `with-redefs`. I suppose that could be an issue in cljs. Is that ok in cljs?

bshepherdson replied:
Ah, good catch. I updated these tests to `^:synchronized`. (Though as you say, it doesn't matter in JS.) I also added `cljs.core/with-redefs` and friends to the linter's list of things disallowed in parallel tests. In `:require` etc. namespace that start with `clojure.*` get rewritten to `cljs.*` but that doesn't happen in the linter.

adam-james-v committed:
Number formatter Reports correct Decimal places for negative Numbers (#46104)

* Number formatter Reports correct Decimal places for negative Numbers

Fixes #41640

Prior to this, the number formatter would get the incorrect number of decimal places for negative numbers.

The way the decimal places count is derived works as follows:

 - convert number to a string
 - split the string on the separator (decimal, comma, whatever the instance has set)
 - destructure that split into [int decimal]
 - count decimal to get the number of decimals.

That all works for positive numbers, but for negative numbers, the regex was just splitting on 'not digits', so we got
an extra group due to the leading negative sign. This would incorrectly destructure  the 'int' part of the number into
the count, thus we would always get 1 instead of the correct value.

This fixes it by just taking last instead of destructuring.

* correct comparison to always use abs value in conditonal check

* tiny incorrectness in the test
